storage/jellyfish_merkle/src/node_type/mod.rs

```diff
@@ -0,0 +1,505 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Node types of [`JellyfishMerkleTree`](crate::JellyfishMerkleTree)
+//!
+//! This module defines two types of Jellyfish Merkle tree nodes: [`InternalNode`],
+//! and [`LeafNode`] as building blocks of a 256-bit
+//! [`JellyfishMerkleTree`](crate::JellyfishMerkleTree). [`InternalNode`] represents a 4-level
+//! binary tree to optimize for IOPS: it compresses a tree with 31 nodes into one node with 16
+//! chidren at the lowest level. [`LeafNode`] stores the full key and the account blob data
+//! associated.
+
+#[cfg(test)]
+mod node_type_test;
+
+use bincode::{deserialize, serialize};
+use crypto::{
+    hash::{
+        CryptoHash, SparseMerkleInternalHasher, SparseMerkleLeafHasher,
+        SPARSE_MERKLE_PLACEHOLDER_HASH,
+    },
+    HashValue,
+};
+use failure::{Fail, Result};
+use serde::{Deserialize, Serialize};
+use sparse_merkle::nibble_path::NibblePath;
+use std::collections::hash_map::HashMap;
+use types::{
+    account_state_blob::AccountStateBlob,
+    proof::{SparseMerkleInternalNode, SparseMerkleLeafNode},
+    transaction::Version,
+};
+
+/// The unique key of each node.
+#[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, Deserialize)]
+pub struct NodeKey {
+    version: Version,
+    nibble_path: NibblePath,
+}
+
+impl NodeKey {
+    /// Creates a new `NodeKey`.
+    pub fn new(version: Version, nibble_path: NibblePath) -> Self {
+        NodeKey {
+            version,
+            nibble_path,
+        }
+    }
+
+    /// Gets the version.
+    pub fn version(&self) -> Version {
+        self.version
+    }
+
+    /// Gets the nibble path.
+    pub fn nibble_path(&self) -> &NibblePath {
+        &self.nibble_path
+    }
+
+    /// Generates a child node_key based on this node key.
+    pub fn gen_child_node_key(&self, version: Version, n: u8) -> NodeKey {
+        let mut node_nibble_path = self.nibble_path().clone();
+        node_nibble_path.push(n);
+        NodeKey::new(version, node_nibble_path)
+    }
+
+    /// Generates a new node key with the same nibble path but a different version.
+    pub fn update_version(&self, version: Version) -> NodeKey {
+        Self {
+            version,
+            nibble_path: self.nibble_path.clone(),
+        }
+    }
+}
+
+/// A shortcut to generate a node key consisting of a version and empty nibble path.
+impl From<Version> for NodeKey {
+    fn from(version: Version) -> Self {
+        Self::new(version, NibblePath::new(vec![]))
+    }
+}
+
+/// Each child of [`InternalNode`] encapsulates a nibble forking at this node.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct Child {
+    // The hash value of this child node.
+    pub hash: HashValue,
```

Comment:
Not very consistent: other structs in this file all have private fields.

Reply:
I feel it's fine to use it like a plain c struct here (with pub fields) since it doesn't have any logic encapsulated.